### PR TITLE
Implement the container app; add helm chart templates; add release automation; add build CI

### PR DESCRIPTION
**Pull request by zliang-akamai**
#### Testing approach
- Apply all manifest to a LKE cluster, `kubectl apply -f /path/to/your/manifests`
- Verify the pod is ok, `kubectl get pods`, you should see one decorator pod for each node.
- Verify the app is running correctly by checking the log, `kubectl logs YOUR_POD_NAME`
- See the result of label updates of nodes, `kubectl get nodes -o wide --show-labels`

#### Generate manifest from Helm chart
- `helm template helm/k8s-node-decorator > my_manifest.yaml`
- `kubectl apply -f my_manifest.yaml`

### CODE REVIEW

File: main.go
Comment (schinmai-akamai) [inline main.go:30]: nit: error message should say `NODE_NAME`

**Review by schinmai-akamai — APPROVED** [review]

File: main.go
Comment (zliang-akamai) [reply main.go]: Thanks, I just fixed it!

File: main.go
Comment (avestuk) [inline main.go:84]: Why do we care if `instanceData` is nil here? 

If we get nil instanceData to begin with we will never update the node labels, which doesn't seem desirable?

**Review by avestuk — APPROVED** [review]

File: Dockerfile
Comment (avestuk) [inline Dockerfile:1]: Can we set a user here so we don't run as root

File: manifest/app.yaml
Comment (avestuk) [inline manifest/app.yaml:1]: Would be great to add a securityContext so we enforce running as non root and set an RO filesystem. Dropping privileges will be the most annoying to do so I won't block on that.

Comment (zliang-akamai) [reply manifest/app.yaml:1]: Hi @avestuk, thanks for the feedback! These manifests in this PR were for testing purposes only, and I am going to create a Helm chart templates for generating manifests that contains much more production-ready configurations like non-root and read only file system and more.

Comment (zliang-akamai) [reply manifest/app.yaml:1]: Actually, I can push it into this PR, later today. When I can verify the generated manifests work.

Comment (avestuk) [reply manifest/app.yaml:1]: Ok great, up to you how it's handled :)

File: main.go
Comment (zliang-akamai) [reply main.go:84]: I think it will be safer to check `nil` of a pointer. I think I can add a warning log into the else branch when the `nil` appears. Do you think that's a good idea?

Comment (avestuk) [reply main.go:84]: As in check that `data` is not nil? I can definitely see the benefit in that, adding an else branch that'll log if it is nil makes sense to me. I guess it's running customer side but even so, if they let support know that there's a lot of `nil` messages being sent on the channel that'd be something for us to look into!

**Review by schinmai-akamai — APPROVED** [review]
Looks great!

File: Dockerfile
Comment (luthermonson) [inline Dockerfile:18]: you sure we need this?

File: main.go
Comment (luthermonson) [inline main.go:134]: why not just pass by reference?

File: main.go
Comment (luthermonson) [inline main.go:118]: `flag.DurationVar` maybe?

File: main.go
Comment (luthermonson) [inline main.go:87]: you could do this in another issue/pr if you'd like but this would be an easy project to get a real context working, maybe start with a timeout context and set it to something < the interval so you never have your watchers exceeding the interval time?

File: Dockerfile
Comment (zliang-akamai) [reply Dockerfile]: I just removed it and it works well. Thanks!

File: main.go
Comment (luthermonson) [inline main.go:28]: a little weird having an `os.Getenv` especially if we move this to a package. maybe better to pass a node name to this func and you could setup a package var you set in main and do the empty check there. this would be much simpler too as just...

```
return clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
```

File: main.go
Comment (luthermonson) [inline main.go:62]: i know the watcher wont emit an event unless the data changed but you should also gate the Update with a diff check as well, no point spamming the Updates if there are no changes to any of the fields. Just if check instanceData to the node.Labels and return nil if nothing has changed should be enough

File: main.go
Comment (luthermonson) [inline main.go:109]: this could lead to a situation where the labels are never updated on the node yet the decorator continues to happily run, e.g. if the kube api errors out and we only log the error we will never receive the Update again from the watcher... some ideas...
1) you could leave this to a best effort... maybe some sort of back off and retry for now?
2) error out all the way and force the pod to restart which would start the update process on boot which does a fresh pull of metadata... 
3) maybe do both? have any better ideas?

File: main.go
Comment (zliang-akamai) [reply main.go]: I marked it as TODO so we can do it in the following up PR. Thanks for the suggestion!

File: main.go
Comment (luthermonson) [inline main.go:89]: do we actually want to continue if we can't use the metadata service? perhaps this is a hard stop unable to boot scenario instead of allowing it to go past and spam a broken metadata service? better to put the pod into crash loop backoff if the metadata service is down?

File: main.go
Comment (zliang-akamai) [reply main.go:134]: Fixed, thanks!

File: main.go
Comment (luthermonson) [inline main.go:144]: just do `klog.Fatal` instead of all these panics?

File: main.go
Comment (luthermonson) [inline main.go:60]: we should probably get a better format than this, something like?

```
decorator.linode.com/label
decorator.linode.com/instance-id
decorator.linode.com/region
decorator.linode.com/instance-type
decorator.linode.com/host
```

File: main.go
Comment (zliang-akamai) [reply main.go:109]: I think just `klog .Fatal` would be fine. I am a little bit worried about infinity restart loop, but I think k8s may have some limit to handle it? like number of restarts allowed during a timeframe?

File: main.go
Comment (luthermonson) [inline main.go:55]: missing the tags labels too? going to be hard to get a `[]string` into a label but i think we should make an attempt in this PR

Comment (schinmai-akamai) [reply main.go:55]: make a csv string?

File: helm/k8s-node-decorator/values.yaml
Comment (luthermonson) [inline helm/k8s-node-decorator/values.yaml:15]: can we get the docker repo made for this before we commit this PR?

File: main.go
Comment (zliang-akamai) [reply main.go:60]: But we haven't set DNS for the domain `decorator.linode.com`? Is it okay to use it here without making the domain point to this repo?

File: main.go
Comment (zliang-akamai) [reply main.go:89]: Sometimes API might have some intermittent errors but I think we can still crash and let k8s handle it's lifecycle to retry the service

File: helm/k8s-node-decorator/values.yaml
Comment (zliang-akamai) [reply helm/k8s-node-decorator/values.yaml:15]: So we are going to build automated release workflow in this PR? Sounds good to me!

File: main.go
Comment (zliang-akamai) [reply main.go:28]: But that will require passing `nodeName` to everywhere. Do you think that would be fine?
```diff
diff --git a/main.go b/main.go
index f403f4c..ae113db 100644
--- a/main.go
+++ b/main.go
@@ -2,7 +2,6 @@ package main
 
 import (
        "context"
-       "errors"
        "flag"
        "fmt"
        "os"
@@ -24,12 +23,7 @@ func init() {
        _ = flag.Set("logtostderr", "true")
 }
 
-func GetCurrentNode(clientset *kubernetes.Clientset) (*corev1.Node, error) {
-       nodeName := os.Getenv("NODE_NAME")
-       if nodeName == "" {
-               return nil, errors.New("Environment variable NODE_NAME is not set")
-       }
-
+func GetCurrentNode(clientset *kubernetes.Clientset, nodeName string) (*corev1.Node, error) {
        node, err := clientset.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{})
        if err != nil {
                return nil, fmt.Errorf("failed to get node: %w", err)
@@ -41,12 +35,13 @@ func GetCurrentNode(clientset *kubernetes.Clientset) (*corev1.Node, error) {
 func UpdateNodeLabels(
        clientset *kubernetes.Clientset,
        instanceData *metadata.InstanceData,
+       nodeName string,
 ) error {
        if instanceData == nil {
                return fmt.Errorf("instance data received from Linode metadata service is nil")
        }
 
-       node, err := GetCurrentNode(clientset)
+       node, err := GetCurrentNode(clientset, nodeName)
        if err != nil {
                return fmt.Errorf("failed to get the node: %w", err)
        }
@@ -83,12 +78,17 @@ func GetClientset() (*kubernetes.Clientset, error) {
        return clientset, nil
 }
 
-func StartDecorator(client metadata.Client, clientset *kubernetes.Clientset, interval time.Duration) {
+func StartDecorator(
+       client metadata.Client,
+       clientset *kubernetes.Clientset,
+       interval time.Duration,
+       nodeName string,
+) {
        instanceData, err := client.GetInstance(context.TODO())
        if err != nil {
                klog.Errorf("Failed to get the initial instance data: %s", err.Error())
        } else {
-               err = UpdateNodeLabels(clientset, instanceData)
+               err = UpdateNodeLabels(clientset, instanceData, nodeName)
        }
 
        if err != nil {
@@ -104,7 +104,7 @@ func StartDecorator(client metadata.Client, clientset *kubernetes.Clientset, int
        for {
                select {
                case data := <-instanceWatcher.Updates:
-                       err = UpdateNodeLabels(clientset, data)
+                       err = UpdateNodeLabels(clientset, data, nodeName)
                        if err != nil {
                                klog.Fatal(err)
                        }
@@ -130,7 +130,12 @@ func main() {
                klog.Fatal(err)
        }
 
-       _, err = GetCurrentNode(clientset)
+       nodeName := os.Getenv("NODE_NAME")
+       if nodeName == "" {
+               klog.Fatal("Environment variable NODE_NAME is not set")
+       }
+
+       _, err = GetCurrentNode(clientset, nodeName)
        if err != nil {
                klog.Fatal(err)
        }
@@ -143,5 +148,5 @@ func main() {
                klog.Fatal(err)
        }
 
-       StartDecorator(*client, clientset, *interval)
+       StartDecorator(*client, clientset, *interval, nodeName)
 }
```

Comment (luthermonson) [reply main.go:28]: make it a package var then and just reference it from that, no point in pulling it from env every time

File: main.go
Comment (luthermonson) [reply main.go]: ya that's a crash loop back off which is fine for kube to just handle. the scheduler will start an exponential backoff after a bit and you wont even notice it

File: main.go
Comment (luthermonson) [reply main.go:60]: ya it's just namespacing things you use... we can call it whatever we want but we especially dont want it to clash with anything else someone installs i their cluster. this is the more formal version of `linode_` as you did, the dns doesnt actually have to work.

File: main.go
Comment (luthermonson) [reply main.go:55]: oof i hate it... no one can target pods to `decorator.linode.com/tags=tag1,tag2,tag3` especially when they're all over the place and out of order? i literally have no better ideas tho

File: helm/k8s-node-decorator/values.yaml
Comment (luthermonson) [reply helm/k8s-node-decorator/values.yaml:15]: let's gooooo 🚀

Comment (luthermonson) [reply helm/k8s-node-decorator/values.yaml:15]: i literally have no idea how to get that repo made tho, gonna need your hep on that lol

File: helm/k8s-node-decorator/.helmignore
Comment (luthermonson) [inline helm/k8s-node-decorator/.helmignore:12]: 😆

**Comment by luthermonson** [comment]
Closes #1

File: main.go
Comment (zliang-akamai) [reply main.go:28]: Oh yea, that's right. Will do

File: main.go
Comment (zliang-akamai) [reply main.go]: I just made it to be `klog.Fatal`. Consider this as solved. Thank you for the suggestion!

File: helm/k8s-node-decorator/.helmignore
Comment (zliang-akamai) [reply helm/k8s-node-decorator/.helmignore:12]: It was automatically generated, lol

File: main.go
Comment (zliang-akamai) [reply main.go:70]: It has been implemented, thanks!

File: main.go
Comment (lgarber-akamai) [inline main.go:98]: nitpick: I think we can drop the `else` here since klog.Fatalf should implicitly exit.

File: helm/k8s-node-decorator/values.yaml
Comment (lgarber-akamai) [reply helm/k8s-node-decorator/values.yaml:15]: https://hub.docker.com/repository/docker/linode/k8s-node-decorator

File: main.go
Comment (lgarber-akamai) [inline main.go:70]: Could a log statement be added after this request is run?

File: main.go
Comment (lgarber-akamai) [inline main.go:64]: If one of the `SetLabel` operations returns true the rest of the operations will be skipped because the first condition will be met for the or.

Maybe we could simplify it down using an anonymous function:

```suggestion
	labelsUpdated := false

	handleUpdated := func(updated bool) {
		if updated {
			labelsUpdated = updated
		}
	}

	handleUpdated(SetLabel(node, "decorator.linode.com/label", instanceData.Label))
	handleUpdated(SetLabel(node, "decorator.linode.com/instance-id", strconv.Itoa(instanceData.ID)))
	handleUpdated(SetLabel(node, "decorator.linode.com/region", instanceData.Region))
	handleUpdated(SetLabel(node, "decorator.linode.com/instance-type", instanceData.Type))
	handleUpdated(SetLabel(node, "decorator.linode.com/host", instanceData.HostUUID))
```

Alternatively these or statements could be reordered to attempt to set a label first 👍

**Review by lgarber-akamai — CHANGES_REQUESTED** [review]
Great work, just a few small things 🙂

File: main.go
Comment (hcwagner) [reply main.go:60]: seems like the only good option here is to request upstream that tags be a kv pair. enumeration would be ok too, but I think given the structure of tags inside Linode that might be more problematic elsewhere. 

without that I think we're probably hosed for lke scheduling based on tags.

Comment (schinmai-akamai) [reply main.go:60]: @luthermonson perhaps we could sort before we add tags? its super annoying but would at least get you consistent ordering..

**Review by lgarber-akamai — APPROVED** [review]
Works well on my end!

Tested with the following:
```bash
export KUBECONFIG=...

docker build --push -t lgarberakamai/k8s-node-decorator:latest .

helm install node-decorator ./helm/k8s-node-decorator --set "decorator.image.repository=docker.io/lgarberakamai/k8s-node-decorator" --set "decorator.image.tag=latest"

kubectl get nodes -o=json | jq -r '.items[].metadata.labels | with_entries( select(.key|contains("decorator")))'
```

File: main.go
Comment (avestuk) [reply main.go:60]: What about if we only sync `key=value` tags?

So for the tags "foo", "bar", "baz=bat" we would only use the "baz=bat" tag and we'd label the node as such?

Comment (luthermonson) [reply main.go:60]: i like this but let's go another level on the "subdomain", `tags.decorator.linode.com/baz=bat`, this avoids any clashing of the existing decorator labels. if the tag is just `baz` then set the value to something expected

`baz=bat` becomes `tags.decorator.linode.com/baz=bat`
`baz` becomes `tags.decorator.linode.com/baz=default`

Comment (avestuk) [reply main.go:60]: Ok I like that too, the chunky question is then what should the default value be? We can just have an empty string "" as the value

baz becomes `tags.decorator.linode.com/baz=""`

Comment (hcwagner) [reply main.go:60]: Yea, luther and I discussed that yesterday. Empty string is fine for a default value and you can split on `=` or `:`

```
tags: ["foo:bat", "bar=", "baz"]

output labels:
tags.decorator.linode.com/foo: "bat"
tags.decorator.linode.com/bar: ""
tags.decorator.linode.com/baz: ""
```

File: helm/k8s-node-decorator/values.yaml
Comment (zliang-akamai) [reply helm/k8s-node-decorator/values.yaml:15]: Just added a release workflow: https://github.com/linode/k8s-node-decorator/pull/6/commits/1dc83774a53234c6a71ae5ed984d50707588da95

I tested with my docker hub account and it works fine:
- https://github.com/zliang-akamai/k8s-node-decorator/actions/runs/7468592765
- https://hub.docker.com/repository/docker/zliangatakamai/k8s-node-decorator/tags

File: main.go
Comment (luthermonson) [inline main.go:141]: after a discussion with the mds team can you bump the default to 5min

File: helm/k8s-node-decorator/values.yaml
Comment (luthermonson) [inline helm/k8s-node-decorator/values.yaml:11]: increase to 5m

File: main.go
Comment (zliang-akamai) [reply main.go:72]: For multiple operators (`:`, or `=`) scenarios, I think we should only respect the first operator:
```
tags: ["foo:bar:baz", "bar=foo=baz", "baz=foo:bar"]

output labels:
tags.decorator.linode.com/foo: "bar:baz"
tags.decorator.linode.com/bar: "foo=baz"
tags.decorator.linode.com/baz: "foo:bar"
```

And for the duplicated keys scenarios, I think we can always use the latter value: 
```
tags: ["foo=bar", "foo=baz"]

output labels:
tags.decorator.linode.com/foo: "baz"
```

Do you guys think this is a good idea?

Comment (lgarber-akamai) [reply main.go:72]: Both of those seem reasonable to me ^

Comment (hcwagner) [reply main.go:72]: I agree with only honoring the first operator, that makes sense to me. Would be interested in other opinions though.

For the duplicated keys I don't think we can honor both/either correctly. AFAIK labels are implemented as a `map[string]string` so we'd have to either honor the first one or the last one. Since we don't have any enforced ordering I think we're going to fall into an area of undefined behavior. I'm thinking we want to report an error, not apply either, and try to propagate it up through the observability stack.

Maybe there's a better way of notifying the user that there's a duplicate?

Comment (avestuk) [reply main.go:72]: I think we can push validation of the map to APINext. I image it can validate that there are no duplicates in the map.

Comment (zliang-akamai) [reply main.go:72]: @hcwagner Ah sorry, I forgot labels were in a map. Thanks for reminding me! Yep, in that case, we can override it with a latter value?

```
tags: ["foo=bar", "foo=baz"]

output labels:
tags.decorator.linode.com/foo: "baz"
```

Comment (avestuk) [reply main.go:72]: @zliang-akamai You're probably right, I suppose the tags are basically `[]string` so you can have multiple instances of the same key with different values. 

Imo what we should do is a stable sort of the tags we get from the API and then set the K:V pairs incrementally. That way we've at least got consistent behaviour.

Comment (hcwagner) [reply main.go:67]: Seems like it's a decent starting point. I still suggest we'll want to provide some visibility into whether there are duplicates at some point. Even though it will consistently end with the same result I definitely see the potential for weirdness occurring in the scheduler when pods are assigned based on affinities set against labels that the decorator flips every time it comes through. It wouldn't be immediately obvious what the problem is with the scheduling if you're debugging from k9s or something like that unless you directly see the flipping of the label value occur.

Comment (zliang-akamai) [reply main.go:81]: Thanks everyone! The tags handling logic has been implemented. Let me know if you have any feedback/suggestion about it, ^_^

File: k8snodedecorator/tags.go
Comment (hcwagner) [inline k8snodedecorator/tags.go:62]: Good idea switching it return a map here. That should cover the possible pain point of scheduling I was worried about.

**Review by hcwagner — APPROVED** [review]
Looks good to me.

**Review by schinmai-akamai — APPROVED** [review]

File: main.go
Comment (lgarber-akamai) [inline main.go:90]: I might be misreading this but it looks like this snippet might not drop deleted tags.

Maybe something like this would work? (untested)
```suggestion
	oldTags := make(map[string]string)
	maps.Copy(node.Labels, oldTags)
	
	newTags := decorator.ParseTags(instanceData.Tags)

	for key, value := range oldTags {
		if !strings.HasPrefix(key, decorator.TagLabelPrefix) {
			continue
		}

		// If the tag is no longer defined, drop it
		if _, ok := newTags[key]; !ok {
			delete(newTags, key)
			labelsUpdated = true
			continue
		}

		// Otherwise attempt to update the tag as normal
		handleUpdated(SetLabel(node, key, value))
	}
```

Comment (zliang-akamai) [reply main.go:90]: Oops, thanks for discovering that! Will fix it

Comment (zliang-akamai) [reply main.go:90]: The suggested change here doesn't seem to be adding new tags? Let me try to fix it based on this

Comment (lgarber-akamai) [reply main.go:90]: Oops you're right, my bad lol

**Review by lgarber-akamai — APPROVED** [review]
Looks perfect, excellent work!

**Review by luthermonson — APPROVED** [review]

**Pull request merged**
